### PR TITLE
Add constructor for ArgumentDef

### LINKED ISSUE #47493: Unable to create TorchScript compatible CUDA extension
## 🐛 Bug

```cuda
// test.cu
#include <torch/extension.h>

bool is_installed() { return true; }

TORCH_LIBRARY(my_ops, m) {
  m.def("is_installed() -> bool", &is_installed);
}
```

```python
# setup.py
from setuptools import setup
import torch
from torch.utils.cpp_extension import CUDAExtension, BuildExtension

setup(
    name='test',
    ext_modules=[
        CUDAExtension('test', ['test.cu'])
    ],
    cmdclass={
        'build_ext': BuildExtension
    }
)
```

Error:
```
[1/1] /usr/local/cuda/bin/nvcc -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/TH -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.8 -c -c /home/gaoxiang/extension-jit/test.cu -o /home/gaoxiang/extension-jit/build/temp.linux-x86_64-3.8/test.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1014"' -DTORCH_EXTENSION_NAME=test -D_GLIBCXX_USE_CXX11_ABI=1 -gencode=arch=compute_86,code=sm_86 -std=c++14
FAILED: /home/gaoxiang/extension-jit/build/temp.linux-x86_64-3.8/test.o 
/usr/local/cuda/bin/nvcc -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/csrc/api/include -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/TH -I/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/THC -I/usr/local/cuda/include -I/usr/include/python3.8 -c -c /home/gaoxiang/extension-jit/test.cu -o /home/gaoxiang/extension-jit/build/temp.linux-x86_64-3.8/test.o -D__CUDA_NO_HALF_OPERATORS__ -D__CUDA_NO_HALF_CONVERSIONS__ -D__CUDA_NO_BFLOAT16_CONVERSIONS__ -D__CUDA_NO_HALF2_OPERATORS__ --expt-relaxed-constexpr --compiler-options ''"'"'-fPIC'"'"'' -DTORCH_API_INCLUDE_EXTENSION_H '-DPYBIND11_COMPILER_TYPE="_gcc"' '-DPYBIND11_STDLIB="_libstdcpp"' '-DPYBIND11_BUILD_ABI="_cxxabi1014"' -DTORCH_EXTENSION_NAME=test -D_GLIBCXX_USE_CXX11_ABI=1 -gencode=arch=compute_86,code=sm_86 -std=c++14
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/boxing/impl/make_boxed_from_unboxed_functor.h(326): warning: variable "num_ivalue_args" was declared but never referenced
          detected during:
            instantiation of "std::decay_t<c10::guts::infer_function_traits<Functor>::type::return_type> c10::impl::call_functor_with_args_from_stack_<Functor,AllowDeprecatedTypes,ivalue_arg_indices...>(Functor *, c10::Stack *, std::index_sequence<ivalue_arg_indices...>) [with Functor=c10::impl::WrapFunctionIntoRuntimeFunctor<std::decay_t<__nv_bool ()>>, AllowDeprecatedTypes=false, ivalue_arg_indices=<>]" 
(346): here
            instantiation of "std::decay_t<c10::guts::infer_function_traits<Functor>::type::return_type> c10::impl::call_functor_with_args_from_stack<Functor,AllowDeprecatedTypes>(Functor *, c10::Stack *) [with Functor=c10::impl::WrapFunctionIntoRuntimeFunctor<std::decay_t<__nv_bool ()>>, AllowDeprecatedTypes=false]" 
(396): here
            instantiation of "void c10::impl::make_boxed_from_unboxed_functor<KernelFunctor, AllowDeprecatedTypes>::call(c10::OperatorKernel *, const c10::OperatorHandle &, c10::Stack *) [with KernelFunctor=c10::impl::WrapFunctionIntoRuntimeFunctor<std::decay_t<__nv_bool ()>>, AllowDeprecatedTypes=false]" 
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/boxing/KernelFunction_impl.h(109): here
            instantiation of "c10::KernelFunction c10::KernelFunction::makeFromUnboxedFunctor<AllowLegacyTypes,KernelFunctor>(std::unique_ptr<c10::OperatorKernel, std::default_delete<c10::OperatorKernel>>) [with AllowLegacyTypes=false, KernelFunctor=c10::impl::WrapFunctionIntoRuntimeFunctor<std::decay_t<__nv_bool ()>>]" 
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/boxing/KernelFunction_impl.h(175): here
            instantiation of "c10::KernelFunction c10::KernelFunction::makeFromUnboxedRuntimeFunction(FuncType *) [with AllowLegacyTypes=false, FuncType=__nv_bool ()]" 
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h(92): here
            instantiation of "torch::CppFunction::CppFunction(Func *, std::enable_if_t<c10::guts::is_function_type<Func>::value, std::nullptr_t>) [with Func=__nv_bool ()]" 
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h(457): here
            instantiation of "torch::Library &torch::Library::def(NameOrSchema &&, Func &&) & [with NameOrSchema=const char (&)[23], Func=__nv_bool (*)()]" 
/home/gaoxiang/extension-jit/test.cu(6): here

/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h: In instantiation of 'constexpr std::array<c10::detail::infer_schema::ArgumentDef, sizeof... (Ts)> c10::detail::infer_schema::createArgumentVectorFromTypes(std::index_sequence<ns ...>) [with Ts = {bool}; long unsigned int ...Is = {0}; std::index_sequence<ns ...> = std::integer_sequence<long unsigned int, 0>]':
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:80:55:   required from 'c10::FunctionSchema c10::detail::infer_schema::createFunctionSchemaFromTraitsFlattenedReturns(std::string&&, std::string&&) [with FunctionTraits = c10::guts::function_traits<bool()>; std::string = std::__cxx11::basic_string<char>]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:146:121:   required from 'c10::FunctionSchema c10::inferFunctionSchemaFlattenedReturns(std::string&&, std::string&&) [with FuncType = bool(); std::string = std::__cxx11::basic_string<char>]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/op_registration.h:23:152:   required from 'std::unique_ptr<c10::FunctionSchema> c10::detail::inferFunctionSchemaFromFunctor() [with KernelFunctor = c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<bool (*)(), bool, c10::guts::typelist::typelist<> >]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h:91:363:   required from 'torch::CppFunction::CppFunction(Func*, std::enable_if_t<c10::guts::is_function_type<FuncType_>::value, std::nullptr_t>) [with Func = bool(); std::enable_if_t<c10::guts::is_function_type<FuncType_>::value, std::nullptr_t> = std::nullptr_t]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h:457:13:   required from 'torch::Library& torch::Library::def(NameOrSchema&&, Func&&) & [with NameOrSchema = const char (&)[23]; Func = bool (*)()]'
/home/gaoxiang/extension-jit/test.cu:6:46:   required from here
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:120:86:   in 'constexpr' expansion of 'c10::detail::infer_schema::createReturns<bool, void>::call()'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:89:55:   in 'constexpr' expansion of 'c10::detail::infer_schema::createReturns<std::tuple<bool>, void>::call()'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:48:82: error: no matching function for call to 'c10::detail::infer_schema::ArgumentDef::ArgumentDef(c10::TypePtr (*)())'
   48 |   return (
      |                                                                                  ^                                                       
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note: candidate: 'c10::detail::infer_schema::ArgumentDef::ArgumentDef()'
   22 | struct ArgumentDef final {
      |        ^~~~~~~~~~~
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note:   candidate expects 0 arguments, 1 provided
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note: candidate: 'constexpr c10::detail::infer_schema::ArgumentDef::ArgumentDef(const c10::detail::infer_schema::ArgumentDef&)'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note:   no known conversion for argument 1 from 'c10::TypePtr (*)()' {aka 'std::shared_ptr<c10::Type> (*)()'} to 'const c10::detail::infer_schema::ArgumentDef&'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note: candidate: 'constexpr c10::detail::infer_schema::ArgumentDef::ArgumentDef(c10::detail::infer_schema::ArgumentDef&&)'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:22:8: note:   no known conversion for argument 1 from 'c10::TypePtr (*)()' {aka 'std::shared_ptr<c10::Type> (*)()'} to 'c10::detail::infer_schema::ArgumentDef&&'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h: In instantiation of 'c10::FunctionSchema c10::detail::infer_schema::createFunctionSchemaFromTraitsFlattenedReturns(std::string&&, std::string&&) [with FunctionTraits = c10::guts::function_traits<bool()>; std::string = std::__cxx11::basic_string<char>]':
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:146:121:   required from 'c10::FunctionSchema c10::inferFunctionSchemaFlattenedReturns(std::string&&, std::string&&) [with FuncType = bool(); std::string = std::__cxx11::basic_string<char>]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/op_registration.h:23:152:   required from 'std::unique_ptr<c10::FunctionSchema> c10::detail::inferFunctionSchemaFromFunctor() [with KernelFunctor = c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<bool (*)(), bool, c10::guts::typelist::typelist<> >]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h:91:363:   required from 'torch::CppFunction::CppFunction(Func*, std::enable_if_t<c10::guts::is_function_type<FuncType_>::value, std::nullptr_t>) [with Func = bool(); std::enable_if_t<c10::guts::is_function_type<FuncType_>::value, std::nullptr_t> = std::nullptr_t]'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/torch/library.h:457:13:   required from 'torch::Library& torch::Library::def(NameOrSchema&&, Func&&) & [with NameOrSchema = const char (&)[23]; Func = bool (*)()]'
/home/gaoxiang/extension-jit/test.cu:6:46:   required from here
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:120:86:   in 'constexpr' expansion of 'c10::detail::infer_schema::createReturns<bool, void>::call()'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:89:55:   in 'constexpr' expansion of 'c10::detail::infer_schema::createReturns<std::tuple<bool>, void>::call()'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:80:55:   in 'constexpr' expansion of 'c10::detail::infer_schema::createArgumentVectorFromTypes<bool, 0>((std::make_index_sequence<1>(), std::make_index_sequence<1>()))'
/home/gaoxiang/.local/lib/python3.8/site-packages/torch/include/ATen/core/op_registration/infer_schema.h:120:16: error: 'constexpr' call flows off the end of the function
  120 |  constexpr auto returns = createReturns<ReturnType>::call();
      |                ^~~~~~~
ninja: build stopped: subcommand failed.
```

## Environment

```
Collecting environment information...
PyTorch version: 1.8.0a0+ceb16d8
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Arch Linux (x86_64)
GCC version: (GCC) 10.2.0
Clang version: 10.0.1 
CMake version: version 3.18.4

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 3090
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 455.28
cuDNN version: Probably one of the following:
/usr/lib/libcudnn.so.8.0.4
/usr/lib/libcudnn_adv_infer.so.8.0.4
/usr/lib/libcudnn_adv_train.so.8.0.4
/usr/lib/libcudnn_cnn_infer.so.8.0.4
/usr/lib/libcudnn_cnn_train.so.8.0.4
/usr/lib/libcudnn_ops_infer.so.8.0.4
/usr/lib/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.18.1
[pip3] pytorch-sphinx-theme==0.0.19
[pip3] torch==1.8.0a0
[pip3] torchani==2.3.dev42+g5a856ee.d20201105
[pip3] torchvision==0.8.0a0+a9c78f1
[conda] Could not collect
```

cc @yf225 @glaringlee @gmagogsfm

### PR DESCRIPTION
**Pull request by zasdfgbnm**
Fixes https://github.com/pytorch/pytorch/issues/47493

### CODE REVIEW

File: aten/src/ATen/core/op_registration/infer_schema.h [hunk 1]
```diff
@@ -22,6 +22,8 @@ namespace infer_schema {
 struct ArgumentDef final {
   using GetTypeFn = TypePtr();
   GetTypeFn* getTypeFn;
+  ArgumentDef() = default;
+  constexpr ArgumentDef(GetTypeFn *getTypeFn): getTypeFn(getTypeFn) {}
```
Comment (dzhulgakov) [inline aten/src/ATen/core/op_registration/infer_schema.h:26]: nit: add `explicit`?

File: aten/src/ATen/core/op_registration/infer_schema.h [hunk 2]
```diff
@@ -22,6 +22,8 @@ namespace infer_schema {
 struct ArgumentDef final {
   using GetTypeFn = TypePtr();
   GetTypeFn* getTypeFn;
+  ArgumentDef() = default;
```
Comment (dzhulgakov) [inline aten/src/ATen/core/op_registration/infer_schema.h:25]: do we actually need a default constructor? if so, I'd set getTypeFn to be nullptr explicitly

**Review by dzhulgakov — APPROVED** [review]

File: aten/src/ATen/core/op_registration/infer_schema.h (hunk 1 above)
Comment (zasdfgbnm) [reply aten/src/ATen/core/op_registration/infer_schema.h]: fixed

File: aten/src/ATen/core/op_registration/infer_schema.h (hunk 2 above)
Comment (zasdfgbnm) [reply aten/src/ATen/core/op_registration/infer_schema.h]: This is only needed by MSVC. Without this, we can not create an `std::array` of `ArgumentDef`

Comment (zasdfgbnm) [reply aten/src/ATen/core/op_registration/infer_schema.h]: I have update this line to set getTypeFn to nullptr

**Comment by zasdfgbnm** [comment]
ping @dzhulgakov

**Pull request closed without merging**